ddtrace/internal/processor.py

```diff
@@ -12,18 +15,76 @@
 
 
 @attr.s
-class TraceProcessor(object):
-    _filters = attr.ib(type=List[TraceFilter])
-
-    def process(self, trace):
-        # type: (List[Span]) -> Optional[List[Span]]
-        for filtr in self._filters:
-            try:
-                trace = filtr.process_trace(trace)
-            except Exception:
-                log.error("error applying filter %r to traces", filtr, exc_info=True)
-            else:
-                if trace is None:
-                    log.debug("dropping trace due to filter %r", filtr)
-                    return
+class Processor(six.with_metaclass(abc.ABCMeta)):
+    """A Processor is used to process spans as they are created and finished by a tracer."""
+
+    def __attrs_post_init__(self):
+        # type () -> None
+        """Default post initializer which logs the representation of the
+        Processor at the ``logging.DEBUG`` level.
+
+        The representation can be modified with the ``repr`` argument to the
+        attrs attribute::
+
+            @attr.s
+            class MyProcessor(Processor):
+                field_to_include = attr.ib(repr=True)
+                field_to_exclude = attr.ib(repr=False)
+        """
+        log.debug("initialized processor %r", self)
+
+    @abc.abstractmethod
+    def on_span_start(self, span):
+        # type: (Span) -> None
+        """Called when a span is started.
+
+        This method is useful for making upfront decisions on spans.
+
+        For example, a sampling decision can be made when the span is created
+        based on its resource name.
+        """
+        pass
+
+    @abc.abstractmethod
+    def on_span_finish(self, data):
+        # type: (Any) -> Any
+        """Called with the result of any previous processors or initially with
+        the finishing span when a span finishes.
+
+        It can return any data which will be passed to any processors that are
+        applied afterwards.
+        """
+        pass
+
+
+@attr.s(eq=False)
+class TraceFilterProcessor(Processor):
+    """Processor for applying a ``TraceFilter`` to traces.
+
+    If a filter raises an exception it will be logged at the ``logging.ERROR``
+    level and the trace will be returned.
+    """
+
+    _filter = attr.ib(type=TraceFilter, repr=True)
```

Comment:
Looks great but why do you need this intermediate class at all if you can do this:

```python
class FilterRequestsOnUrl(Processor):
    def on_span_finish(self, trace):
       # type: (List[Span]) -> Optional[List[Span]]
       for span in trace:
            if span.parent_id is None and span.get_tag(http.URL) is not None:
                url = span.get_tag(http.URL)
                for regexp in self._regexps:
                    if regexp.match(url):
                        return None
        return trace
```

Reply:
We definitely should longer term, we can't right now because it's in the public API: https://ddtrace.readthedocs.io/en/stable/advanced_usage.html#trace-filtering

Reply:
Ah, that's a good answer. 😅  I guess that's fine for now then.